test/forest_test.cpp

```diff
@@ -0,0 +1,290 @@
+/**************************************************************************************************/
+
+// stdc++
+#include <iostream>
+
+// boost
+#include <boost/test/unit_test.hpp>
+
+// stlab
+#include <stlab/forest.hpp>
+
+/**************************************************************************************************/
+
+using namespace stlab;
+
+/**************************************************************************************************/
+
+BOOST_AUTO_TEST_CASE(empty_forest) {
+    forest<int> f;
+    BOOST_CHECK(f.empty());
+    BOOST_CHECK(f.size() == 0);
+    BOOST_CHECK(f.begin() == f.end());
+}
+
+/**************************************************************************************************/
+
+BOOST_AUTO_TEST_CASE(single_node_forest) {
+    forest<int> f;
+
+    auto il = f.insert(f.end(), 42);
+    BOOST_CHECK(il.edge() == forest_edge::leading);
+    BOOST_CHECK(*il == 42);
+    BOOST_CHECK(!f.empty());
+
+    BOOST_CHECK(f.begin() != f.end());
+    BOOST_CHECK(f.size() == 1);
+
+    auto it = trailing_of(il);
+    BOOST_CHECK(it.edge() == forest_edge::trailing);
+    BOOST_CHECK(*it == *il);
+}
+
+/**************************************************************************************************/
+
+namespace {
+
+/**************************************************************************************************/
+
+auto big_test_forest() {
+    forest<std::string> f;
+
+    auto a_iter = trailing_of(f.insert(f.end(), "A"));
+
+    auto b_iter = trailing_of(f.insert(a_iter, "B"));
+
+    auto c_iter = trailing_of(f.insert(b_iter, "C"));
+    auto d_iter = trailing_of(f.insert(b_iter, "D"));
+    auto e_iter = f.insert(b_iter, "E");
+
+    f.insert(c_iter, "F");
+    f.insert(c_iter, "G");
+    f.insert(c_iter, "H");
+
+    f.insert(d_iter, "I");
+    f.insert(d_iter, "J");
+    f.insert(d_iter, "K");
+
+    BOOST_CHECK(f.size() == 11);
+
+    BOOST_CHECK(has_children(a_iter));
+    BOOST_CHECK(has_children(b_iter));
+    BOOST_CHECK(has_children(c_iter));
+    BOOST_CHECK(has_children(d_iter));
+    BOOST_CHECK(!has_children(e_iter));
+
+    return f;
+}
+
+/**************************************************************************************************/
+
+template <typename R>
+inline auto to_string(R&& r) {
+    std::string result;
+    for (const auto& x : r) {
+        result += x;
+    }
+    return result;
+}
+
+template <typename I>
+inline auto to_string(I first, I last) {
+    std::string result;
+    while (first != last) {
+        result += *first++;
+    }
+    return result;
+}
+
+/**************************************************************************************************/
+
+template <typename Iterator>
+void test_fullorder_traversal(Iterator first, Iterator last, std::string expected) {
```

Comment:
Is there a reason that expected is not a const&? I don't see a reason for a copy.

Reply:
Agreed & changed.